crates/sui-graphql-rpc/src/mutations.rs

```diff
@@ -0,0 +1,100 @@
+use sui_types::{signature::GenericSignature, transaction::Transaction};
+
+pub struct Mutation;
+
+#[Object]
+impl Mutation {
+    async fn execute_transaction_block(
+        &self,
+        ctx: &Context<'_>,
+        tx_bytes: String,
+        signatures: Vec<String>,
+    ) -> Result<ExecutionResult> {
+        let sui_sdk_client: &SuiClient = ctx.data().map_err(|_| {
+            graphql_error(
+                code::INTERNAL_SERVER_ERROR,
+                "Unable to fetch Sui SDK client",
+            )
+        })?;
+
+        let tx_data = bcs::from_bytes(
+            &Base64::try_from(tx_bytes)
+                .map_err(|e| {
+                    graphql_error(
+                        code::INTERNAL_SERVER_ERROR,
+                        format!("Unable to deserialize transaction to Base64: {:?}", e),
+                    )
+                })?
+                .to_vec()
+                .map_err(|e| {
+                    graphql_error(
+                        code::INTERNAL_SERVER_ERROR,
+                        format!("Unable to decode Base64 to vec: {:?}", e),
+                    )
+                })?,
+        )?;
+
+        let mut sigs = Vec::new();
+        for sig in signatures {
+            sigs.push(
+                GenericSignature::from_bytes(
+                    &Base64::try_from(sig)
+                        .map_err(|e| {
+                            graphql_error(
+                                code::INTERNAL_SERVER_ERROR,
+                                format!("Unable to deserialize signature to Base64: {:?}", e),
+                            )
+                        })?
+                        .to_vec()
+                        .map_err(|e| {
+                            graphql_error(
+                                code::INTERNAL_SERVER_ERROR,
+                                format!("Unable to decode Base64 to vec: {:?}", e),
+                            )
+                        })?,
+                )
+                .map_err(|e| {
+                    graphql_error(
+                        code::INTERNAL_SERVER_ERROR,
+                        format!("Unable to create generic signature from bytes: {:?}", e),
+                    )
+                })?,
+            );
+        }
+        let transaction =
+            Transaction::from_generic_sig_data(tx_data, Intent::sui_transaction(), sigs);
+
+        let result = sui_sdk_client
+            .quorum_driver_api()
+            .execute_transaction_block(
+                transaction,
+                SuiTransactionBlockResponseOptions::default(),
+                Some(ExecuteTransactionRequestType::WaitForEffectsCert),
+            )
+            .await
+            .map_err(|e| {
+                graphql_error(
+                    code::INTERNAL_SERVER_ERROR,
+                    format!("Unable to execute transaction: {:?}", e),
+                )
+            })?;
+
+        Ok(ExecutionResult {
+            effects: None,
```

Comment:
We cannot actually get effects here. The RPC throws an error that local exec must be enabled for effects to be returned.
Hence we should remove this field.
cc @amnn @stefan-mysten

Reply:
Copying out-of-band response:

We can't get rid of the effects field entirely, unfortunately, because without it, we introduce indexer post-processing latency into the execution inner loop.  Although the fullnode does not have enough information to return a full effects response, it should have enough to produce an `EffectsCertificate` which has the raw transaction effect.  That is what we need.

We shouldn't need to change the schema drastically to support this, but we will need to make some changes to its representation (that's partly what motivated the change in representation to `TransactionBlockEffects` here: https://github.com/MystenLabs/sui/pull/15038):

- `TransactionBlockEffects` (after the PR above) is represented by its stored (indexed) form and its native form.  To support outputting effects from transactions that have just executed, we need to make that stored form optional.
- Further -- we need to support a form of Object that is just an object reference (i.e. Object becomes an enum that is either the fully formed object with its contents, or just an ID, version and digest).

The reason that a lot of fields are optional is largely to support these kinds of use cases where the same GraphQL type is created in different circumstances and so has access to different pieces of information.

Reply:
On further digging, I don't see how we can get `EffectsCertificate` from the FN RPC response without doing an RPC call to the FN after the TX exec.
The response of TX exec is [SuiTransactionBlockResponse](https://github.com/MystenLabs/sui/blob/1265c0ada4dd75d49f90bb65614716f33746bc51/crates/sui-json-rpc-types/src/sui_transaction.rs#L189) and it doesn't contain anything like `EffectsCertificate`. What am I missing here?

Reply:
When fullnode executes a transaction, it has access to the `EffectsCertificate` here:

https://github.com/MystenLabs/sui/blob/c774f8452d6d5f428e9101d9a8ae531aabae6ac0/crates/sui-json-rpc/src/transaction_execution_api.rs#L157

It converts it into the `effects` field of `SuiTransactionBlockResponse` here:

https://github.com/MystenLabs/sui/blob/c774f8452d6d5f428e9101d9a8ae531aabae6ac0/crates/sui-json-rpc/src/transaction_execution_api.rs#L202

AFAICT, this should work even if local execution is not enabled, because that requirement only applies for computing object changes and balance changes.  In the case of object changes, after #15095, they will no longer be required, because they will be calculated from the `TransactionEffects` directly, and for balance changes, it will mean that we do not offer balance changes on a transaction that has not been indexed (where the `stored` field is `None`).

So this field should contain all the right information, just in the wrong type (you want `TransactionEffects` which is the type of the `native` field in GraphQL's `TransactionBlockEffects`, and you got `SuiTransactionBlockEffects` which is the JSON representation of that type).